### ISSUE TITLE
New build system: how much is it compatible with Debian ?

### ISSUE THREAD

**Issue by lopippo** [OpenMS/OpenMS#4717]
Greetings, 
I wanted to build OpenMS package for Debian using our own Debian ways. I discovered the new build system that was crafted around the real-time download of a number of libraries, their patching and then their building. First of all, that really seems to work fine, as I have almost finished building the contrib libs (Debian testing).
In Debian, we try to avoid using external library code in the source tree. We would rather use the system-packaged libraries. I checked the versions of the downloaded libraries against Debian unstable ones and they match almost cipher-to-cipher. 
Question: do you think that the build system could be adapted to use the sytem libraries? If so, would the missing patches applied by OpenMS be critical for the functioning of the programs ? I checked some patch instances and some apply modifs for MacOSX, for instance. Are there patches that actually fix bugs in the libraries ?
Thanks for listening,
Filippo

Edit: after looking at the various binaries that are built (libs and execs) with objdump | grep NEEDED, I see that the only libs needed are system libs, like 

  NEEDED               libOpenMS.so
  NEEDED               libOpenSwathAlgo.so
  NEEDED               libQt5Core.so.5
  NEEDED               libQt5Gui.so.5
  NEEDED               libQt5Network.so.5
  NEEDED               libQt5Svg.so.5
  NEEDED               libQt5Widgets.so.5
  NEEDED               libc.so.6
  NEEDED               libdl.so.2
  NEEDED               libgcc_s.so.1
  NEEDED               libgomp.so.1
  NEEDED               libm.so.6
  NEEDED               libpthread.so.0
  NEEDED               libstdc++.so.6
  NEEDED               libz.so.1

So I gather that the external libs mentioned above are used with static linking ? Is this true or am I misunderstanding something ?

Thanks for your help

Filippo

**Comment by jpfeuffer** [comment]
Hey Filippo!

Although the contrib build system is not really new, I think we made it more prominent in the documentation by now such that users on older/exotic OSes have less trouble finding all the right versions (and added the OPENMS_CONTRIB_LIBS variable for users that are not so familiar with CMake).
However, we did not intend to recommend this as the "first choice" for users. Actually, the suggested way is still to try to build it with the shared libraries from the system and if and only if some version is missing or clashes, to resort to the contrib build system.
(The contrib build system indeed builds static by default [for our standalone packages of the executables] but can be switched to shared libraries if requested).

I am relatively confident, too, that all or most library versions in the latest Debian should be
compatible. Without patches.

TLDR: try to skip the contrib build and hope that CMake finds all the system libraries. Adapt the CMAKE_PREFIX_PATH if needed, as usual.

**Comment by lopippo** [comment]
Greetings, Julianus !

Thank you very much for your answer. Indeed, the tools/quickbuild.sh script automates all this, and the CMake art that you collectively master (I'm impressed, really) takes the user to the final product as if nothing were going on!
I am very much relieved that the "use the shared libs provided by the system, if possible" philosophy is still your main way of handling things. I have just finished configuring OpenMS and the build has started. That is very good software engineering, very much appreciated for such a comprehensive software package!

I hope that my packaging of 2.5.0 will be easy and quick. When the package is done, I'll inform you and close this issue. Let's keep it open for discussion while I package 2.5.0, in case I have questions, if you do not mind.

Cheers,
Filippo

**Comment by lopippo** [comment]
Hello, 
I am glad to tell that I could build all the libraries and binaries just fine. However, I have a problem with the documentation. While the LaTeX output seems to be correctly generated, its build into a PDF file is blocked by the error below that I cannot solve. And I am stuck at this point... (Follows below).

8< ~~~~~~~~~
~/devel/packaging/openms/development/debian/build/doc/OpenMS_tutorial/latex_output$ make
rm -f *.ps *.dvi *.aux *.toc *.idx *.ind *.ilg *.log *.out *.brf *.blg *.bbl refman.pdf
pdflatex refman
This is pdfTeX, Version 3.14159265-2.6-1.40.20 (TeX Live 2019/Debian) (preloaded format=pdflatex)
 restricted \write18 enabled.
entering extended mode
(./refman.tex
LaTeX2e <2020-02-02> patch level 5
L3 programming layer <2020-02-25>
(/usr/share/texlive/texmf-dist/tex/latex/base/article.cls
Document Class: article 2019/12/20 v1.4l Standard LaTeX document class
(/usr/share/texlive/texmf-dist/tex/latex/base/size10.clo))
(/usr/share/texlive/texmf-dist/tex/latex/geometry/geometry.sty
(/usr/share/texlive/texmf-dist/tex/latex/graphics/keyval.sty)
(/usr/share/texlive/texmf-dist/tex/generic/iftex/ifvtex.sty
(/usr/share/texlive/texmf-dist/tex/generic/iftex/iftex.sty)))
(/usr/share/texlive/texmf-dist/tex/latex/base/makeidx.sty)
(/usr/share/texlive/texmf-dist/tex/latex/fancyhdr/fancyhdr.sty)
(/usr/share/texlive/texmf-dist/tex/latex/graphics/graphicx.sty
(/usr/share/texlive/texmf-dist/tex/latex/graphics/graphics.sty
(/usr/share/texlive/texmf-dist/tex/latex/graphics/trig.sty)
(/usr/share/texlive/texmf-dist/tex/latex/graphics-cfg/graphics.cfg)
(/usr/share/texlive/texmf-dist/tex/latex/graphics-def/pdftex.def)))
(/usr/share/texlive/texmf-dist/tex/latex/tools/multicol.sty)
(/usr/share/texlive/texmf-dist/tex/latex/float/float.sty)
(/usr/share/texlive/texmf-dist/tex/latex/base/alltt.sty)
(/usr/share/texlive/texmf-dist/tex/latex/newtx/newtxtext.sty
`newtxtext' v1.531, 2018/03/27 Text macros taking advantage of TeX-Gyre Termes 
fonts (msharpe) (/usr/share/texlive/texmf-dist/tex/latex/fontaxes/fontaxes.sty)
(/usr/share/texlive/texmf-dist/tex/latex/xkeyval/xkeyval.sty
(/usr/share/texlive/texmf-dist/tex/generic/xkeyval/xkeyval.tex
(/usr/share/texlive/texmf-dist/tex/generic/xkeyval/xkvutils.tex)))
(/usr/share/texlive/texmf-dist/tex/latex/etoolbox/etoolbox.sty)
(/usr/share/texlive/texmf-dist/tex/latex/base/fontenc.sty
(/usr/share/texlive/texmf-dist/tex/latex/newtx/t1ntxtlf.fd))
(/usr/share/texlive/texmf-dist/tex/latex/base/textcomp.sty))
(/usr/share/texlive/texmf-dist/tex/latex/newtx/newtxmath.sty
`newtxmath' v1.624, 2020/01/25 Math macros based originally on txfonts (msharpe
) (/usr/share/texlive/texmf-dist/tex/latex/amsmath/amsmath.sty
For additional information on amsmath, use the `?' option.
(/usr/share/texlive/texmf-dist/tex/latex/amsmath/amstext.sty
(/usr/share/texlive/texmf-dist/tex/latex/amsmath/amsgen.sty))
(/usr/share/texlive/texmf-dist/tex/latex/amsmath/amsbsy.sty)
(/usr/share/texlive/texmf-dist/tex/latex/amsmath/amsopn.sty))
(/usr/share/texlive/texmf-dist/tex/generic/iftex/ifxetex.sty)
(/usr/share/texlive/texmf-dist/tex/generic/iftex/ifluatex.sty)
(/usr/share/texlive/texmf-dist/tex/latex/oberdiek/centernot.sty)
(/usr/share/texlive/texmf-dist/tex/generic/kastrup/binhex.tex))
(/usr/share/texlive/texmf-dist/tex/latex/hyperref/hyperref.sty
(/usr/share/texlive/texmf-dist/tex/generic/ltxcmds/ltxcmds.sty)
(/usr/share/texlive/texmf-dist/tex/latex/pdftexcmds/pdftexcmds.sty
(/usr/share/texlive/texmf-dist/tex/generic/infwarerr/infwarerr.sty))
(/usr/share/texlive/texmf-dist/tex/generic/kvsetkeys/kvsetkeys.sty)
(/usr/share/texlive/texmf-dist/tex/generic/kvdefinekeys/kvdefinekeys.sty)
(/usr/share/texlive/texmf-dist/tex/generic/pdfescape/pdfescape.sty)
(/usr/share/texlive/texmf-dist/tex/latex/hycolor/hycolor.sty)
(/usr/share/texlive/texmf-dist/tex/latex/letltxmacro/letltxmacro.sty)
(/usr/share/texlive/texmf-dist/tex/latex/auxhook/auxhook.sty)
(/usr/share/texlive/texmf-dist/tex/latex/kvoptions/kvoptions.sty)
(/usr/share/texlive/texmf-dist/tex/latex/hyperref/pd1enc.def)
(/usr/share/texlive/texmf-dist/tex/generic/intcalc/intcalc.sty)
(/usr/share/texlive/texmf-dist/tex/generic/etexcmds/etexcmds.sty)
(/usr/share/texlive/texmf-dist/tex/latex/hyperref/backref.sty
(/usr/share/texlive/texmf-dist/tex/latex/rerunfilecheck/rerunfilecheck.sty
(/usr/share/texlive/texmf-dist/tex/latex/atveryend/atveryend.sty)
(/usr/share/texlive/texmf-dist/tex/generic/uniquecounter/uniquecounter.sty
(/usr/share/texlive/texmf-dist/tex/generic/bigintcalc/bigintcalc.sty))))
(/usr/share/texlive/texmf-dist/tex/latex/url/url.sty)
(/usr/share/texlive/texmf-dist/tex/generic/bitset/bitset.sty)
(/usr/share/texlive/texmf-dist/tex/generic/atbegshi/atbegshi.sty))
(/usr/share/texlive/texmf-dist/tex/latex/hyperref/hpdftex.def) (./doxygen.sty
(/usr/share/texlive/texmf-dist/tex/latex/tools/calc.sty)
(/usr/share/texlive/texmf-dist/tex/latex/tools/verbatim.sty)
(/usr/share/texlive/texmf-dist/tex/latex/xcolor/xcolor.sty
(/usr/share/texlive/texmf-dist/tex/latex/graphics-cfg/color.cfg)
(/usr/share/texlive/texmf-dist/tex/latex/colortbl/colortbl.sty
(/usr/share/texlive/texmf-dist/tex/latex/tools/array.sty)))
(./longtable_doxygen.sty) (./tabu_doxygen.sty
(/usr/share/texlive/texmf-dist/tex/latex/varwidth/varwidth.sty))
(/usr/share/texlive/texmf-dist/tex/latex/fancyvrb/fancyvrb.sty)
(/usr/share/texlive/texmf-dist/tex/latex/tools/tabularx.sty)
(/usr/share/texlive/texmf-dist/tex/latex/multirow/multirow.sty)
(/usr/share/texlive/texmf-dist/tex/latex/hanging/hanging.sty)
(/usr/share/texlive/texmf-dist/tex/generic/iftex/ifpdf.sty)
(/usr/share/texlive/texmf-dist/tex/latex/adjustbox/adjustbox.sty
(/usr/share/texlive/texmf-dist/tex/latex/adjustbox/adjcalc.sty)
(/usr/share/texlive/texmf-dist/tex/latex/adjustbox/trimclip.sty
(/usr/share/texlive/texmf-dist/tex/latex/collectbox/collectbox.sty)
(/usr/share/texlive/texmf-dist/tex/latex/adjustbox/tc-pdftex.def))
(/usr/share/texlive/texmf-dist/tex/latex/ifoddpage/ifoddpage.sty))
(/usr/share/texlive/texmf-dist/tex/latex/amsfonts/amssymb.sty
(/usr/share/texlive/texmf-dist/tex/latex/amsfonts/amsfonts.sty)

! LaTeX Error: Command `\Bbbk' already defined.

See the LaTeX manual or LaTeX Companion for explanation.
Type  H <return>  for immediate help.
 ...                                              
                                                  
l.261 ...ol{\Bbbk}           {\mathord}{AMSb}{"7C}
                                                  
?

~~~~~~~~~ >8

I searched for Bbbk in the texlive distribution code and found the occurrences below, but I cannot go anywhere from there... Any help ?

Ciao
Filippo

/usr/share/texlive/texmf-dist/tex$ grep -i 'Bbbk' * -r
fontinst/mathetx/msbm.etx:   \comment{The symbol `$\Bbbk$'.}
generic/tex4ht/mathml.4ht:\:temp \Bbbk
generic/ofs/ofs-tx.tex:  \mathchardef\varBbbk      "0\hex\txmiafam AB
generic/ofs/ofs-ams.tex:  \mathchardef\Bbbk              "0\hex\msbmfam 7C
latex/stix/stix.sty:\stix@MathSymbol{\Bbbk}                     {\mathord}{symbols3}{`k}
latex/txfonts/txfonts.sty:\re@DeclareMathSymbol{\Bbbk}{\mathord}{AMSb}{124}
latex/txfonts/txfonts.sty:\DeclareMathSymbol{\varBbbk}{\mathord}{lettersA}{171}
latex/newpx/newpxmath.sty:\re@DeclareMathSymbol{\Bbbk}{\mathord}{AMSm}{124}
latex/newtx/newtxmath.sty:\re@DeclareMathSymbol{\Bbbk}{\mathord}{AMSm}{107} % was 124
latex/newtx/newtxmath.sty:    \iftx@ams\else\let\mathbb\vmathbb \let\Bbbk=\vmathbb{k}\fi
latex/newtx/newtxmath.sty:\DeclareRobustCommand{\Bbbk}{\mathbb{k}}
latex/mathdesign/mdpgd/mdpgd.sty:    \MD@def\DeclareMathSymbol{\Bbbk}           {\mathord}{mathdesignA}{"7C}
latex/mathdesign/mdput/mdput.sty:    \MD@def\DeclareMathSymbol{\Bbbk}           {\mathord}{mathdesignA}{"7C}
latex/mathdesign/mdugm/mdugm.sty:    \MD@def\DeclareMathSymbol{\Bbbk}           {\mathord}{mathdesignA}{"7C}
latex/mathdesign/mdpus/mdpus.sty:    \MD@def\DeclareMathSymbol{\Bbbk}           {\mathord}{mathdesignA}{"7C}
latex/mathdesign/mdici/mdici.sty:    \MD@def\DeclareMathSymbol{\Bbbk}           {\mathord}{mathdesignA}{"7C}
latex/mathdesign/mdbch/mdbch.sty:    \MD@def\DeclareMathSymbol{\Bbbk}           {\mathord}{mathdesignA}{"7C}
latex/stix2/stix2.sty:\stix@MathSymbol{\Bbbk}                     {\mathord}{symbols3}{`k}
latex/mnsymbol/MnSymbol.sty:  \let\Bbbk\undefined
latex/pxfonts/pxfonts.sty:\re@DeclareMathSymbol{\Bbbk}{\mathord}{AMSb}{124}
latex/typeface/typeface.sty:    \tf@undefinecommands{\Bbbk} % prevent the clash of glyphs
latex/elegantbook/elegantbook.cls:  \let\Bbbk\relax
latex/elegantbook/elegantbook.cls:  \let\Bbbk\relax
latex/amsfonts/amssymb.sty:\DeclareMathSymbol{\Bbbk}           {\mathord}{AMSb}{"7C}
latex/boisik/boisik.sty:\DeclareMathSymbol{\Bbbk}                 {\mathord}{bsksymbols}{"6B}
latex/newtxsf/newtxsf.sty:\re@DeclareMathSymbol{\Bbbk}{\mathord}{AMSm}{124}
latex/fdsymbol/fdsymbol.sty:\let\Bbbk\undefined
latex/fdsymbol/fdsymbol.sty:\newcommand{\Bbbk}{\mathbb{\mathchar"717C}}
latex/ncctools/nccbbb.sty:\newcommand\bbbk{\NCC@bbb{K}}
latex/breqn/msabm.sym:\DeclareFlexSymbol{\Bbbk}            {Ord}{MSB}{7C}
latex/kpfonts/kpfonts.sty:\re@DeclareMathSymbol{\Bbbk}{\mathord}{AMSb}{124}
latex/arev/ams-mdbch.sty:    \DeclareMathSymbol{\Bbbk}           {\mathord}{mathdesignA}{"7C}
latex/unicode-math/unicode-math-table.tex:\UnicodeMathSymbol{"1D542}{\BbbK                     }{\mathalpha}{mathematical double-struck capital k}%
latex/unicode-math/unicode-math-table.tex:\UnicodeMathSymbol{"1D55C}{\Bbbk                     }{\mathalpha}{mathematical double-struck small k}%
plain/amsfonts/amssym.tex:\newsymbol\Bbbk 207C

**Comment by lopippo** [comment]
Greetings,

I had to patch the libwildmagic make files to force linking against the other project's libraries even if no symbols were used. Inded, some binaries from OpenMS did use constructors not used during the wild magic libs building. The new libwildmagic package will be uploaded soon. Further, that is the 5p17 version (and last one, as the libs are no more developed). Are you considering using the Geometric Tools new libs in the future (https://www.geometrictools.com/) ?
Filippo

**Comment by jpfeuffer** [comment]
Hi!

Regarding the latex, doxygen issue: Which doxygen version are you using. There is an incompatibility of doxygen <=1.8.15 with newer Latex distributions (in particular of the tabu package): https://github.com/doxygen/doxygen/issues/6769
We had big problems at the time this happened as well. There should be a Debian bug report already. We build our own, new doxygen now. In between, we actually had some patch somewhere that replaced a ".sty" on the build machines after setup (https://github.com/OpenMS/build-scripts/blob/357a4db7896e2cdf1ce14323c52cbca75633c3e4/platformScripts/shell/linux/installLatex.sh#L24).

Regarding WildMagic, see this issue #4581 . Yes, with some small adaptions we should be able to use the last WM5 version (as it seems you have figured out already). I just didn't get to check it on all platforms yet, so we did not include it in the contrib distribution yet. And yes, I am confident that the few parts where we using WildMagic probably would work with GeometricTools as well but since it only affects a small part of our library the priority is not so high yet.

**Comment by lopippo** [comment]
Greetings, Julianus,

thank you for your answer. I could manage to make the docs, eventually, with the doc_minimal target.
I finished packaging 2.5.0. Thank you for your patience.

Filippo
